numpy/core/src/multiarray/arrayobject.c

```diff
@@ -387,27 +440,24 @@ array_dealloc(PyArrayObject *self)
         PyObject_ClearWeakRefs((PyObject *)self);
     }
     if (fa->base) {
-        /*
-         * UPDATEIFCOPY means that base points to an
-         * array that should be updated with the contents
-         * of this array upon destruction.
-         * fa->base->flags must have been WRITEABLE
-         * (checked previously) and it was locked here
-         * thus, unlock it.
-         */
-        if (fa->flags & NPY_ARRAY_UPDATEIFCOPY) {
-            PyArray_ENABLEFLAGS(((PyArrayObject *)fa->base),
-                                                    NPY_ARRAY_WRITEABLE);
-            Py_INCREF(self); /* hold on to self in next call */
-            if (PyArray_CopyAnyInto((PyArrayObject *)fa->base, self) < 0) {
-                PyErr_Print();
-                PyErr_Clear();
-            }
-            /*
-             * Don't need to DECREF -- because we are deleting
-             *self already...
+        int retval;
+        Py_INCREF(self); /* hold on to self in next call */
+        retval = PyArray_ResolveUpdateIfCopy(self);
+        if (retval < 0)
+        {
+            PyErr_Print();
+            PyErr_Clear();
+        }
+#ifdef DEPRECATE_UPDATEIFCOPY
+        if (retval > 0) {
+            DEPRECATE("UPDATEIFCOPY resolution in array_dealloc is "
+                      "incompatible with PyPy and will be removed in "
+                      "the future");
+            /* dealloc must not raise an error, even if warning filters are set
              */
+            PyErr_Clear();
```

Comment:
I think this should use `PyErr_WriteUnraisable()` or similar – otherwise when someone sets a warning filter to deprecations into errors, then the effect will be to hide this warning entirely!